File: src/runtime/zig/std/pathlib_native.zig
// Native implementation of pytra.std.pathlib for Zig.
//
// Provides Path type that works with Zig's []const u8 strings.
// This avoids the PyObject = i64 limitation for string-based types.
//
// source: src/pytra/std/pathlib.py

const std = @import("std");
const pytra = @import("../built_in/py_runtime.zig");
const os_native = @import("os_native.zig");
const os_path_native = @import("os_path_native.zig");

pub const Path = struct {
    _value: []const u8,

    pub fn init(value: anytype) Path {
        return Path{ ._value = pytra.union_as_str(value) };
    }

    pub fn __str__(self: *const Path) []const u8 {
        return self._value;
    }

    pub fn __repr__(self: *const Path) []const u8 {
        _ = self;
        return "Path(...)";
    }

    pub fn __fspath__(self: *const Path) []const u8 {
        return self._value;
    }

    pub fn __truediv__(self: *const Path, rhs: []const u8) *Path {
        return pytra.make_object(Path, Path.init(os_path_native.join(self._value, rhs)));
    }

    pub fn parent(self: *const Path) *Path {
        const d = os_path_native.dirname(self._value);
        const result = if (d.len == 0) "." else d;
        return pytra.make_object(Path, Path.init(result));
    }

    pub fn name(self: *const Path) []const u8 {
        return os_path_native.basename(self._value);
    }

    pub fn suffix(self: *const Path) []const u8 {
        const bn = os_path_native.basename(self._value);
        if (std.mem.lastIndexOfScalar(u8, bn, '.')) |dot| {
            return bn[dot..];
        }
        return "";
    }

    pub fn stem(self: *const Path) []const u8 {
        const bn = os_path_native.basename(self._value);
        if (std.mem.lastIndexOfScalar(u8, bn, '.')) |dot| {
            return bn[0..dot];
        }
        return bn;
    }

    pub fn exists(self: *const Path) bool {
        return os_path_native.exists(self._value);
    }

    pub fn mkdir(self: *const Path) void {
        os_native.makedirs(self._value, true);
    }

    pub fn read_text(self: *const Path) []const u8 {
        const file = std.fs.cwd().openFile(self._value, .{}) catch return "";
        defer file.close();
        return file.readToEndAlloc(std.heap.page_allocator, 1 << 20) catch "";
    }

    pub fn write_text(self: *const Path, text: []const u8) void {
        const f = std.fs.cwd().createFile(self._value, .{}) catch return;
        defer f.close();
        f.writeAll(text) catch {};
    }

    pub fn resolve(self: *const Path) *Path {
        return pytra.make_object(Path, Path.init(self._value));
    }

    pub fn joinpath(self: *const Path, rhs: []const u8) *Path {
        return pytra.make_object(Path, Path.init(os_path_native.join(self._value, rhs)));
    }
};

File: src/runtime/nim/std/pathlib_native.nim
import std/os as std_os
import py_runtime

proc Path*(text: string): PyPath =
  text

proc Path*(value: PyObj): PyPath =
  py_str(value)

proc cwd*(): PyPath =
  std_os.getCurrentDir()

proc parent*(path: PyPath): PyPath =
  std_os.parentDir(path)

proc resolve*(path: PyPath): PyPath =
  std_os.absolutePath(path)

proc parents*(path: PyPath): seq[PyPath] =
  var current = std_os.parentDir(std_os.absolutePath(path))
  while current != "" and current != std_os.parentDir(current):
    result.add(current)
    current = std_os.parentDir(current)

proc name*(path: PyPath): string =
  std_os.extractFilename(path)

proc stem*(path: PyPath): string =
  let (_, base, _) = std_os.splitFile(path)
  base

proc mkdir*(path: PyPath): void =
  std_os.createDir(path)

proc mkdir*(path: PyPath, parents: bool, exist_ok: bool): void =
  discard parents
  discard exist_ok
  std_os.createDir(path)

proc joinpath*(path: PyPath, child: string, more: varargs[string]): PyPath =
  result = std_os.joinPath(path, child)
  for part in more:
    result = std_os.joinPath(result, part)

proc `/`*(path: PyPath, child: string): PyPath =
  std_os.joinPath(path, child)

proc exists*(path: PyPath): bool =
  std_os.fileExists(path) or std_os.dirExists(path)

proc write_text*(path: PyPath, text: string): int =
  writeFile(path, text)
  text.len

proc write_text*(path: PyPath, text: string, encoding: string): int =
  discard encoding
  writeFile(path, text)
  text.len

proc read_text*(path: PyPath): string =
  readFile(path)

proc read_text*(path: PyPath, encoding: string): string =
  discard encoding
  readFile(path)

proc read_text*(path: PyPath, encoding: string, errors: string): string =
  discard encoding
  discard errors
  readFile(path)
